ValkeyJSON.md

```diff
@@ -0,0 +1,894 @@
+| $.store.*                                                     | all members of the store                    |
+| $["store"].*                                                  | all members of the store                    |
+| $.store..price                                                | the price of everything in the store        |
+| $..*                                                          | all recursive members of the JSON structure |
+| $..book[*]                                                    | all books                                   |
+| $..book[0]                                                    | the first book                              |
+| $..book[-1]                                                   | the last book                               |
+| $..book[0:2]                                                  | the first two books                         |
+| $..book[0,1]                                                  | the first two books                         |
+| $..book[0:4]                                                  | books from index 0 to 3 (ending index is not inclusive) |
+| $..book[0:4:2]                                                | books at index 0, 2                         |
+| $..book[?(@.isbn)]                                            | all books with isbn number                  |
+| $..book[?(@.price<10)]                                        | all books cheaper than $10                  |
+| '$..book[?(@.price < 10)]'                                    | all books cheaper than $10. (The path must be quoted if it contains whitespaces) |
+| '$..book[?(@["price"] < 10)]'                                 | all books cheaper than $10             |
+| '$..book[?(@.["price"] < 10)]'                                | all books cheaper than $10             |
+| $..book[?(@.price>=10&&@.price<=100)]                         | all books in the price range of $10 to $100, inclusive |
+| '$..book[?(@.price>=10 && @.price<=100)]'                     | all books in the price range of $10 to $100, inclusive. (The path must be quoted if it contains whitespaces) |
+| $..book[?(@.sold==true&#124;&#124;@.in-stock==false)]         | all books sold or out of stock                         |
+| '$..book[?(@.sold == true &#124;&#124; @.in-stock == false)]' | all books sold or out of stock. (The path must be quoted if it contains whitespaces)                 |
+| '$.store.book[?(@.["category"] == "fiction")]'                | all books in the fiction category |
+| '$.store.book[?(@.["category"] != "fiction")]'                | all books in non-fiction categories |
+
+### JSON Command API
+
+The API is compatible with RedisJSON v2. Note that API compatibility here means our command API is a superset of RedisJSON API. 
+For example, we have command “JSON.DEBUG DEPTH” and “JSON.DEBUG FIELDS”, while they do not.
```

Comment:
I would like to know why not have two separate commands, JSON.DEPTH and JSON.FIELDs ? Having a debug command doesn't really make that much sense.

Reply:
The intent is to avoid diverging from RedisJSON unnecessarily. If we introduce new commands, they won't be supported by the existing OSS RedisJSON client libraries. If we have our own ValkeyJSON client library ecosystem, then it would make sense to have them as new commands, instead of sub-commands of JSON.DEBUG.

Reply:
In the RFC, “JSON.DEBUG DEPTH” and “JSON.DEBUG FIELDS” are the only extension. We consider "depth of a json tree/sub-tree" and "number of fields in a json tree/sub-tree" as debug informations. Hence, they are subcommands of JSON.DEBUG. However, it doesn't mean in the future, all extensions should be subcommands of JSON.DEBUG. I am open to adding new JSON commands in the future if necessary.

Reply:
I tried to look up the list of commands elsewhere and my best source for the RedisJSON v2 API didn't agree with this.  Probably I am not looking in the right place.  Please point to it for me.

I saw JSON.DEBUG without arguments (which is not in the PR), and JSON.DEBUG MEMORY (which is), but not JSON.DEBUG HELP (an extension?).  I also saw JSON.MSET (an oversight?).

Reply:
> I also saw JSON.MSET (an oversight?)

Yeah, we missed this. This was added sort of randomly by Redis at some point.

Reply:
RedisJSON API is evolving. Missed JSON.MSET. Will add it to the RFC when I get a chance.

Interestingly, their documentation doesn't seem to match latest docker image `redis/redis-stack:latest`. I don't see I don't see json.mset:
```
docker run -p 6379:6379 redis/redis-stack:latest
...
% ./redis-cli command |grep -i json
json.arrtrim
json.arrpop
json.clear
json.arrappend
json.objlen
json.nummultby
json.mget
json.strappend
json.toggle
json.get
json.arrlen
json.strlen
json.numpowby
json.resp
json.type
json.forget
json.arrindex
json.arrinsert
json.set
json.del
json.numincrby
json.debug
json.objkeys
```

Also, using the same docker image, I only see `JSON.DEBUG HELP` and `JSON.DEBUG` is invalid.
```
127.0.0.1:6379> json.debug help
1) "MEMORY <key> [path] - reports memory usage"
2) "HELP                - this message"
127.0.0.1:6379> json.debug
(error) ERR wrong number of arguments for 'json.debug' command
```

Reply:
Just realized RedisJSON added `json.numpowby` at some point. Will add this to the RFC as well.

Reply:
I have a feeling it might be deprecated. I found out the json.nummultiply is deprecated as per     https://github.com/redis/redis-py/blob/1cbb97ecd7a0bf87aec332a257b589919385d90d/redis/commands/json/commands.py#L144C6-L144C25.